### ISSUE TITLE
[java] CloseResource false-positive on Maven

### ISSUE THREAD

**Issue by bohni** [pmd/pmd#2337]
**Affects PMD Version:** 
6.21.0

**Rule:**
[CloseResource](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#closeresource)

**Description:**
Apache POI POIXMLDocument#close ist not recognized
 
**Code Sample demonstrating the issue:**

see [github repo](https://github.com/bohni/pmd-error)

```java
package com.example.pmd;

import java.io.IOException;

import org.apache.poi.xssf.usermodel.XSSFWorkbook;

public class CloseResourceBug
{

	public void falsePositive() throws IOException
	{
		XSSFWorkbook workbook = new XSSFWorkbook();

		workbook.close();
	}
}

```

**Running PMD through:** *Maven*
```xml
<?xml version="1.0" encoding="UTF-8"?>
<pmd xmlns="http://pmd.sourceforge.net/report/2.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://pmd.sourceforge.net/report/2.0.0 http://pmd.sourceforge.net/report_2_0_0.xsd"
    version="6.21.0" timestamp="2020-03-06T11:28:08.878">
<file name="pmd-error\src\main\java\com\example\pmd\CloseResourceBug.java">
<violation beginline="12" endline="12" begincolumn="30" endcolumn="37" rule="CloseResource" ruleset="Error Prone" package="com.example.pmd" class="CloseResourceBug" method="falsePositive" variable="workbook" externalInfoUrl="https://pmd.github.io/pmd-6.21.0/pmd_rules_java_errorprone.html#closeresource" priority="3">
Ensure that resources like this XSSFWorkbook object are closed after use
</violation>
</file>
</pmd>


```
**Running PMD through:** *cli*
No error reported

```console
$ pmd.bat -d src\main\java\ -f text  -R src\main\config\pmd\pmd-ruleset.xml
Mär 06, 2020 11:33:59 AM net.sourceforge.pmd.PMD encourageToUseIncrementalAnalysis
WARNUNG: This analysis could be faster, please consider using Incremental Analysis: https://pmd.github.io/pmd-6.21.0/pmd_userdocs_incremental_analysis.html

```

$ systeminfo
[...]
Betriebssystemname:                            Microsoft Windows 10 Enterprise
Betriebssystemversion:                         10.0.17763 Nicht zutreffend Build 17763
[...]

$ java -version
openjdk version "1.8.0_232"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_232-b09)
OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.232-b09, mixed mode)

**Comment by adangel** [comment]
Is the violation still reported, if you do the close inside a finally block or use try-with-resources?
Example:

```java
package com.example.pmd;

import java.io.IOException;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;

public class CloseResourceBug {
	public void falsePositive() throws IOException {
		XSSFWorkbook workbook = new XSSFWorkbook();
		try {
			// do stuff, which could raise a runtime exception
		} finally {
			workbook.close();
		}

		// alternative with try-with-resources
		try (XSSFWorkbook w2 = new XSSFWorkbook()) {
			// do stuff, which could raise a runtime exception
		}
	}
}
```

Just simply calling close won't fix the problem - in case of a exception, the method is left before the close call, leaving the workbook open...

**Comment by bohni** [comment]
OK, my fault. Your code doesn't raise the violation.

So the error is in cli, that does not report the error...

**Comment by adangel** [comment]
> So the error is in cli, that does not report the violation...

PMD needs to understand, which type XSSFWorkbook actually is - that is e.g. implements AutoClosable. Therefore we do type resolution and for that, we need access to the full classpath of the project.

With maven, the maven-pmd-plugin determines the classpath automatically (with the help of maven): all dependencies are there.

With CLI, you need to do that on your own. You'll need to use the same classpath as you used for compiling and provide this to PMD via the `-auxclasspath <classpath>` option (https://pmd.github.io/latest/pmd_userdocs_installation.html#running-pmd-via-command-line and https://pmd.github.io/latest/pmd_userdocs_cli_reference.html)

E.g.

```console
$ pmd.bat -d src\main\java\ -f text  -R src\main\config\pmd\pmd-ruleset.xml^
    -auxclasspath target\classes;%USER_HOME%\.m2\repository\org\apache\poi\poi-ooxml\3.6\poi-ooxml-3.6.jar
```

**Comment by bohni** [comment]
OK, I see... RTFM...

**Comment by adangel** [comment]
This is another example why we need #194
